### PR TITLE
Add support for a named POSIX shared memory to OS memory provider

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Add support for a named POSIX shared memory to OS memory provider.

Add a new mechanism for the shared memory mapping to OS provider: a named POSIX shared memory object
The already existing mechanism (an anonymous file descriptor) will be used if a name of a POSIX shared memory object is not provided.

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

**Comment by vinser52** [comment]
Maybe instead of replacing the current implementation we can add additional implementation?
UMF can choose automatically which one to use or user can control it via config option or environment variable.

**Comment by ldorau** [comment]
> Maybe instead of replacing the current implementation we can add additional implementation? UMF can choose automatically which one to use or user can control it via config option or environment variable.

OK, I can do that.

**Comment by ldorau** [comment]
> Maybe instead of replacing the current implementation we can add additional implementation? UMF can choose automatically which one to use or user can control it via config option or environment variable.

@vinser52 @bratpiorka Done. Ready for review.

File: include/umf/providers/provider_os_memory.h [hunk 1]
```diff
@@ -103,10 +105,11 @@ umfOsMemoryProviderParamsDefault(void) {
     umf_os_memory_provider_params_t params = {
         UMF_PROTECTION_READ | UMF_PROTECTION_WRITE, /* protection */
         UMF_MEM_MAP_PRIVATE,                        /* visibility mode */
-        NULL,                                       /* numa_list */
-        0,                                          /* numa_list_len */
-        UMF_NUMA_MODE_DEFAULT,                      /* numa_mode */
-        0                                           /* part_size */
+        NULL,                  /* name of a shared memory file */
```
Comment (bratpiorka) [inline include/umf/providers/provider_os_memory.h:108]: please add an info that this is an optional parameter

File: src/provider/provider_os_memory_linux.c [hunk 2]
```diff
@@ -28,6 +31,21 @@ umf_result_t os_translate_mem_visibility_flag(umf_memory_visibility_t in_flag,
...
 }
 
+// open or create a shared memory file
+int os_shm_open(const char *shm_name) {
+    if (shm_name == NULL) {
+        LOG_ERR("empty name of a shared memory file");
+        return -1;
+    }
+
+    int fd = shm_open(shm_name, O_RDWR, 0600);
```
Comment (bratpiorka) [inline src/provider/provider_os_memory_linux.c:41]: you wrote that the shared memory file must start with a ‘/’ character - could you check this here?

**Review by bratpiorka — CHANGES_REQUESTED** [review]

**Review by kswiecicki — APPROVED** [review]

File: src/provider/provider_os_memory_linux.c (hunk 2 above)
Comment (ldorau) [reply src/provider/provider_os_memory_linux.c:41]: It is checked inside `shm_open()` - see `shm_open(3)`:
` EINVAL The name argument to shm_open() was invalid.`

File: include/umf/providers/provider_os_memory.h (hunk 1 above)
Comment (ldorau) [reply include/umf/providers/provider_os_memory.h]: Done

File: README.md [hunk 3]
```diff
@@ -142,11 +142,26 @@ More detailed documentation is available here: https://oneapi-src.github.io/unif
...
+OS memory provider supports two types of memory mappings (set by the `visibility` parameter):
 1) private memory mapping (`UMF_MEM_MAP_PRIVATE`)
 2) shared memory mapping (`UMF_MEM_MAP_SHARED` - supported on Linux only yet)
 
-If the shared memory mapping is used then an anonymous file descriptor for memory mapping is created using:
+There are available two mechanisms for the shared memory mapping:
+1) a named POSIX shared memory object (used if the `shm_name` parameter is not NULL) or
+2) an anonymous file descriptor (used if the `shm_name` parameter is NULL)
+
+A named POSIX shared memory object is used only if the `shm_name` parameter is not NULL and an anonymous file descriptor is used otherwise.
```
[the remark this answers is not in the document]
Comment (ldorau) [reply README.md]: Done

**Review by bratpiorka — APPROVED** [review]

**Comment by ldorau** [comment]
Rebased

File: README.md [hunk 4]
```diff
@@ -142,11 +142,26 @@ More detailed documentation is available here: https://oneapi-src.github.io/unif
...
 2) shared memory mapping (`UMF_MEM_MAP_SHARED` - supported on Linux only yet)
 
-If the shared memory mapping is used then an anonymous file descriptor for memory mapping is created using:
+There are available two mechanisms for the shared memory mapping:
+1) a named POSIX shared memory object (used if the `shm_name` parameter is not NULL) or
+2) an anonymous file descriptor (used if the `shm_name` parameter is NULL)
+
+A named POSIX shared memory object is used only if the `shm_name` parameter is not NULL, otherwise an anonymous file descriptor is used.
+
+A name of a POSIX shared memory object should be of the form `/somename`; that is, a null-terminated string of up to NAME_MAX (i.e., 255) characters consisting of an initial slash, followed by one or more characters, none of which are slashes (see the shm_open(3) manual). The POSIX shared memory object should be created by a user as a file `/dev/shm/somename` with the appropriate permissions and size in the following way:
```
Comment (vinser52) [inline README.md:156]: > (see the shm_open(3) manual)

UMF is a cross-platform solution. I think we should not refer to POSIX API.

Comment (vinser52) [inline README.md:156]: Why do we require an initial slash in the name? We can add it by ourselves.

Comment (vinser52) [inline README.md:156]: > The POSIX shared memory object should be created by a user as a file `/dev/shm/somename` with the appropriate permissions and size

Why does the user need to create a file?

File: src/provider/provider_os_memory.c [hunk 5]
```diff
@@ -248,35 +248,66 @@ static umf_result_t translate_params(umf_os_memory_provider_params_t *in_params,
...
 
-    provider->fd = os_create_anonymous_fd(provider->visibility);
-    if (provider->fd == -1) {
-        LOG_ERR(
-            "creating an anonymous file descriptor for memory mapping failed");
-        return UMF_RESULT_ERROR_UNKNOWN;
-    }
+    // size_fd will be increased during each allocation if (provider->fd > 0)
+    provider->size_fd = 0;
+    provider->shm_name[0] = 0; // zero shm_name
```
Comment (vinser52) [inline src/provider/provider_os_memory.c:253]: nit: `provider->shm_name[0] = '\0';`

File: src/provider/provider_os_memory.c [hunk 6]
```diff
@@ -248,35 +248,66 @@ static umf_result_t translate_params(umf_os_memory_provider_params_t *in_params,
...
+
+    if (in_params->visibility != UMF_MEM_MAP_SHARED) {
+        provider->fd = -1;
+        provider->max_size_fd = 0;
+    } else if (in_params->shm_name) {
+        provider->fd = os_shm_open(in_params->shm_name);
```
Comment (vinser52) [inline src/provider/provider_os_memory.c:259]: I would refactor this function due to the following concerns:

- The function is too long now. Need to put some code blocks into separate functions.
- The name of the function `translate_params` is misleading. This function not only translates parameters but also open shared memory file or creates anonymous FD, etc.

File: src/provider/provider_os_memory.c [hunk 7]
```diff
@@ -794,6 +837,7 @@ static umf_result_t os_allocation_merge(void *provider, void *lowPtr,
 
 typedef struct os_ipc_data_t {
     int pid;
+    char shm_name[MAX_NAME];
```
Comment (vinser52) [inline src/provider/provider_os_memory.c:840]: You are extending `os_ipc_data_t` by `MAX_NAME` bytes, but it is needed only if named shared memory is used. Perhaps we need two IPC data structures for anonymous and named shared memory.

Comment (vinser52) [inline src/provider/provider_os_memory.c:840]: You are extending `os_ipc_data_t` by `MAX_NAME` bytes, but it is needed only if named shared memory is used. Perhaps we need two IPC data structures for anonymous and named shared memory.

**Review by vinser52 — CHANGES_REQUESTED** [review]

File: src/provider/provider_os_memory.c (hunk 6 above)
Comment (ldorau) [reply src/provider/provider_os_memory.c:248]: Done

File: src/provider/provider_os_memory.c [hunk 8]
```diff
@@ -248,35 +248,66 @@ static umf_result_t translate_params(umf_os_memory_provider_params_t *in_params,
...
+        provider->fd = os_shm_open(in_params->shm_name);
+        if (provider->fd <= 0) {
+            LOG_ERR("opening a shared memory file (/dev/shm%s) for memory "
+                    "mapping failed",
+                    in_params->shm_name);
+            return UMF_RESULT_ERROR_UNKNOWN;
+        }
+
+        strncpy(provider->shm_name, in_params->shm_name, MAX_NAME - 1);
+        provider->shm_name[MAX_NAME - 1] = 0;
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_os_memory.c]: Done

File: src/provider/provider_os_memory.c (hunk 5 above)
Comment (ldorau) [reply src/provider/provider_os_memory.c]: Done

File: README.md (hunk 4 above)
Comment (ldorau) [reply README.md]: It is a standard name of a shared memory object required by `shm_open`.

File: README.md [hunk 9]
```diff
@@ -142,11 +142,26 @@ More detailed documentation is available here: https://oneapi-src.github.io/unif
...
+OS memory provider supports two types of memory mappings (set by the `visibility` parameter):
 1) private memory mapping (`UMF_MEM_MAP_PRIVATE`)
 2) shared memory mapping (`UMF_MEM_MAP_SHARED` - supported on Linux only yet)
 
-If the shared memory mapping is used then an anonymous file descriptor for memory mapping is created using:
+There are available two mechanisms for the shared memory mapping:
+1) a named POSIX shared memory object (used if the `shm_name` parameter is not NULL) or
+2) an anonymous file descriptor (used if the `shm_name` parameter is NULL)
+
+A named POSIX shared memory object is used only if the `shm_name` parameter is not NULL, otherwise an anonymous file descriptor is used.
```
[the remark this answers is not in the document]
Comment (ldorau) [reply README.md]: Removed.

File: src/provider/provider_os_memory_linux.c (hunk 2 above)
Comment (bratpiorka) [inline src/provider/provider_os_memory_linux.c:41]: Could you try passing the O_RDWR|O_CREAT|O_EXCL flags? Maybe that would help with the initial ftruncate?

File: src/provider/provider_os_memory.c (hunk 7 above)
Comment (ldorau) [reply src/provider/provider_os_memory.c]: Done

Comment (ldorau) [reply src/provider/provider_os_memory.c]: Done

**Comment by ldorau** [comment]
Converted to a draft because of a potential TOCTOU issue (`shm_open()`) - needs more investigation.

File: README.md (hunk 4 above)
Comment (ldorau) [reply README.md]: I added an option to create an SHM file if open() fails.
Done.

File: src/provider/provider_os_memory_linux.c (hunk 2 above)
Comment (ldorau) [reply src/provider/provider_os_memory_linux.c:66]: Yes, With the `O_EXCL` flag it does work at last :-)

I added an option to create an SHM file if open() fails.

Done.

File: src/provider/provider_os_memory.c (hunk 7 above)
Comment (vinser52) [reply src/provider/provider_os_memory.c]: Please look into this: https://www.geeksforgeeks.org/flexible-array-members-structure-c/

the common approach is not to specify the array size. And the `os_get_ipc_handle_size` function should be changed accordingly:
```
if (os_provider->shm_name[0]) {
        // os_ipc_data_t->shm_name will be used
        *size = sizeof(os_ipc_data_t) + MAX_NAME;
    } else {
        // os_ipc_data_t->shm_name will NOT be used
        *size = sizeof(os_ipc_data_t);
    }
```

**Comment by ldorau** [comment]
Rebased

File: src/provider/provider_os_memory.c (hunk 7 above)
Comment (ldorau) [reply src/provider/provider_os_memory.c]: @vinser52 I know that. This approach is used when a length of the array is variable. However it is constant in our case, so IMHO the current approach is better in our case.

**Comment by ldorau** [comment]
> Converted to a draft because of a potential TOCTOU issue (`shm_open()`) - it needs more investigation.

We always create a new SHM file with required permissions at once:
```c
shm_open(shm_name, O_RDWR | O_CREAT | O_EXCL, 0600);
```

**Comment by vinser52** [comment]
> @vinser52 please resolve your issues

Done.

File: README.md (hunk 4 above)
Comment (ldorau) [reply README.md]: Done

**Comment by ldorau** [comment]
@vinser52 I fixed the last issue: https://github.com/oneapi-src/unified-memory-framework/pull/506#discussion_r1624370549 - please re-review.

**Review by vinser52 — APPROVED** [review]

**Pull request merged**